pkg/gateway/gateway_test.go

```diff
@@ -4,11 +4,174 @@ package gateway
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			baseLog := logrus.NewEntry(logrus.StandardLogger())
+			db := mock_database.NewMockGateway(controller)
+			metrics := mock_metrics.NewMockEmitter(controller)
+
+			httpl, _ := utilnet.Listen("tcp", ":8080", SocketSize)
+			httpsl, _ := utilnet.Listen("tcp", ":8443", SocketSize)
+
+			env := mock_env.NewMockCore(controller)
+			tt.mocks(env)
+
+			gtwy, err := NewGateway(ctx, env, baseLog, baseLog, db, httpsl, httpl, tt.acrResourceID, tt.gatewayDomains, metrics)
+
+			if tt.wantErr != "" {
+				if err == nil {
+					t.Errorf("Expected error: '%s' but did not receive any errors", tt.wantErr)
+				}
+				if err.Error() != tt.wantErr {
+					t.Errorf("The expected error: '%s' did not match the actual error: '%s'", tt.wantErr, err.Error())
+				}
+				return // no need to continue processing the tests
+			} else if err != nil {
+				t.Errorf("Got an unexpected error: %s", err.Error())
+				return
+			}
+
+			gtwyValue, _ := gtwy.(*gateway)
+
+			if len(gtwyValue.allowList) != len(tt.expectedAllowList) {
+				t.Errorf("expected %d but got %d domains in the allow list", len(tt.expectedAllowList), len(gtwyValue.allowList))
+			}
+			for _, expectedDomain := range tt.expectedAllowList {
+				_, ok := gtwyValue.allowList[expectedDomain]
+				if !ok {
+					t.Errorf("wanted domain: %s but it was missing", expectedDomain)
+				}
+			}
+
+			if gtwyValue.httpl.(*proxyproto.Listener).Listener != httpl {
+				t.Error("httpl was not set properly")
+			}
+			if gtwyValue.httpsl.(*proxyproto.Listener).Listener != httpsl {
+				t.Error("httpsl was not set properly")
+			}
+			if gtwyValue.s.ReadTimeout != 10*time.Second {
+				t.Errorf("gateway http server timeout set to %v when it should be 10s", gtwyValue.s.ReadTimeout)
+			}
+			if gtwyValue.s.IdleTimeout != 2*time.Minute {
+				t.Errorf("gateway http server timeout set to %v when it should be 2m0s", gtwyValue.s.IdleTimeout)
+			}
+			if gtwyValue.s.ErrorLog.Prefix() != "" {
+				t.Errorf("gateway http server ErrorLog should have empty prefix but was found with: %s", gtwyValue.s.ErrorLog.Prefix())
+			}
+			if gtwyValue.s.ErrorLog.Flags() != 0 {
+				t.Errorf("gateway http server ErrorLog should have flags set as 0 but was found with: %d", gtwyValue.s.ErrorLog.Flags())
+			}
+			if gtwyValue.s.BaseContext(httpl) != ctx {
+				t.Error("gateway http server BaseContext should return the ctx of the gateway")
+			}
+			if gtwyValue.s.Handler == nil {
+				t.Error("gateway http server should have its handler set")
+			}
+			if len(gtwyValue.gateways) != 0 {
+				t.Error("gateways should initially be empty")
+			}
+			if gtwyValue.m != metrics {
+				t.Error("metrics was not set properly")
+			}
+			if gtwyValue.ready.Load() != true {
+				t.Error("a new gateway should start as ready")
+			}
```

Comment:
These feel like they should be individual, named tests instead of part of the `for` cases, or is it important to test these default behaviors on every iteration?

Reply:
I don't have any objection to making these individual tests, I'll work on a refactor.